javaagent-extension-api/src/main/java/io/opentelemetry/javaagent/extension/instrumentation/internal/EarlyInstrumentationModule.java

```diff
@@ -0,0 +1,23 @@
+/*
+ * Copyright The OpenTelemetry Authors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package io.opentelemetry.javaagent.extension.instrumentation.internal;
+
+import io.opentelemetry.javaagent.extension.instrumentation.InstrumentationModule;
+import io.opentelemetry.sdk.autoconfigure.spi.Ordered;
+
+/**
+ * Marker interface for instrumentation modules whose virtual fields should be set up before
+ * OpenTelemetry SDK is initialized.
+ *
+ * <p>This class is internal and is hence not for public use. Its APIs are unstable and can change
+ * at any time.
+ */
+public interface EarlyInstrumentationModule extends Ordered {
```

Comment:
Are there any things an `EarlyInstrumentationModule` must not do in it's constructor / static initializers? E.g. logging maybe?

If yes, it would be great to add those restrictions to the javadoc.

Reply:
Since we are not fully set up when we start looking for `EarlyInstrumentationModule` instances there definitely will be things, such as reading configuration, that won't work. In this PR I moved calling `ExperimentalConfig.get()` out of the static initializer of `InstrumentationModule`. In my opinion it is too hard to predict what exactly can or can not be called. I would deal with these issues as they come up. There is a good chance there only the executor instrumentation will use `EarlyInstrumentationModule`.